### ISSUE TITLE
Training: not reflecting

### ISSUE THREAD

**Issue by askshi** [wit-ai/wit#881]
**Possible bug**
Training is not reflecting. I am trying to train the agent with entities which I added already with keywords and synonyms. Its happening since last wednesday.
I already sent a mail to help@wit.ai regarding the issue with subject: Training issue for new phrases
**App ID:** 59916605-fcbb-4fd0-9e58-d19ee895e51c
Possible reference: [Issue](https://github.com/wit-ai/wit/issues/879)

**Comment by patapizza** [comment]
@askshi What is the issue? Can you share a few examples with what is expected vs what is returned?

Thanks.

**Comment by askshi** [comment]
Please find attached url of steps to reproduce the issue.
[](https://drive.google.com/file/d/14H4rEsJUmzz2ho8LERYhB2zHITviy-a2/view)

**Comment by patapizza** [comment]
@askshi Thanks for the screencast.

It seems like you'd like `skin_concern` to recognize certain values, with different synonyms. In this case, the correct strategy is `keywords`, not `free-text` as currently set. If you'd like it to understand not only the list of provided values, but you're open to new ones, then `free-text & keywords` would be the correct one.

Feel free to reopen if updating the entity type doesn't help.

**Comment by askshi** [comment]
I would like to recognize skin_concern and keyword both. And tried with both and both individually. Apart from skin_concern the other entities are also giving same result despite being both keyword and free text. .

**Comment by askshi** [comment]
@patapizza As noted above we've tried with both strategies, `keywords` and `free-text & keywords`, however, we're not able to get any phrase (which has keywords shared with other entities) to be valid against a single entity of our choice, ex. `skin_concern`. You can refer to the following screenshots to see the issue: [Snapshot 1](https://drive.google.com/open?id=1rhj_v2ERMAtv9cu1S56TD8bh9_dXooQW), [Snapshot 2](https://drive.google.com/open?id=1eHfSiDyU5UaGvn9PTb0oYXOO3pMPESfW).

We're facing an overall training issue as well in our agents. Here's a sample with our Chinese agent, where only short phrases with no space or punctuation, seem to be getting trained: [Snapshot 3](https://drive.google.com/file/d/1Hf4Yg6lAQxPDLcSlT8RwTuWx29dptcwc/view?usp=sharing)

**Comment by askshi** [comment]
@patapizza Gentle remainder on the issue.

**Comment by patapizza** [comment]
@askshi Sorry I missed this. Will get another look this week.

**Comment by askshi** [comment]
@patapizza, a gentle reminder. Any update on the issue?

**Comment by patapizza** [comment]
Sorry I got swamped. It's still on my list, I haven't lost sight of it. Thanks for your patience!

**Comment by askshi** [comment]
@patapizza It would be really helpful if we can resolve this as soon as possible.

**Comment by patapizza** [comment]
Hi @askshi,

I've updated the search strategy for `skin_concern` to `free-text & keywords`.

Here are the 2 examples:
![skin break out](https://i.imgur.com/BvkaIH6.png)
![whitehead acne](https://i.imgur.com/kvFPtD8.png)

Is this what you were expecting?
Note that as the value `Acne` as both `whitehead` and `acne` as synonyms, you get 2 `Acne` entities.

Re the Chinese agent, can you open up a new ticket with the app-id and misbehaving inputs with expected vs actual (in text, as a screencast is difficult to repro)?

**Comment by askshi** [comment]
@patapizza The whole word "whitehead acne" is also a synonym. What you are saying is 
![wit ai shi bot singapre](https://user-images.githubusercontent.com/30517297/34596656-9e1b8510-f207-11e7-9478-b62037ba9a8f.png)

Which is correct when Whitehead and acne used separately in the sentences. Since Either one of them or both of them can be used. It can be like : 
1. Show me whitehead
2. Show me acne
3. Show me whitehead acne
In the third case, it should be whitehead acne whole world, which I am trying to train it.
To see another example, lets take another example of dry skin, which is single keyword under skin_type but its getting as : 
![wit ai shi bot singapre 2](https://user-images.githubusercontent.com/30517297/34596783-7e0e18b8-f208-11e7-876c-53f9c99933be.png)
The dataset: 
![wit ai shi bot singapre 3](https://user-images.githubusercontent.com/30517297/34596810-c36516be-f208-11e7-8619-10593e0382d8.png)
Same with "greasy skin":
I hope this will help.

**Comment by patapizza** [comment]
@askshi Got it.

It seems like `product_area` is meant to classify at the sentence level. Have you tried using a `trait` entity for this one (you'd need to validate more than 1 sample)? That way it would let `skin_type` take the full span.

Regarding `whitehead acne`, it seems that this is an edge case where the classifier would favor multiple entities over a single, longer one. In both cases, you'd get an `Acne` value in the API response. Can you provide me more context into why it's important to get a single entity instead of two?

Thanks.

**Comment by askshi** [comment]
@patapizza  product_area will come under sentence, not as a single sentence, so the free text or keyword entity must be used to extract the word from the sentence. and coming to whitehead acne, the whole word consist the acne synonym, likewise the "dry skin" word. We need the whole word as an entity synonym. Acne is a different word, Whitehead is different and whitehead acne is different.

**Comment by patapizza** [comment]
@askshi I'm sorry I don't understand your explanation on `product_area`. Can you reformulate or provide more context/examples? I understand the issue with `dry skin` as it would return 2 different entities instead of one (as a synonym for `Dry`). However I still don't get why returning `whitehead acne` as one entity is important, as `whitehead`, `acne` and `whitehead acne` are all synonyms for `Acne`.

**Comment by askshi** [comment]
@patapizza  both cases required same attention, on the bot part we will extract entity from the API response, we need two entities in response, one is intent and another one is a parameter, our backend has products in Whitehead, acne and whitehead acne and all are different. So we can only throw whitehead acne products. 
Same for dry skin, we need single entity in that keyword because we need to show result based on that entity, it cannot be nested because the keyword being one. On the bot part, the handling of whitehead acne and dry skin will be same, only the entity will be changed.

**Comment by patapizza** [comment]
@askshi What is the name of the entity used as intent detection? `whitehead`, `acne` and `whitehead acne` would all return an entity `skin_concern` with value `Acne`. How do you disambiguate between the products?

**Comment by askshi** [comment]
@patapizza We have included some more examples where the search method recommendations fail our requirements:
*combination-skin*: returns `skin_type` and `product_area` (instead of only `skin_type`)
*normal-skin*: same as above
*greasy skin*: returns `skin_concern` and `product_type` (instead of only `skin_type`)

**Comment by askshi** [comment]
@patapizza  please respond the above. The referenced issues are closed hence it is not helpful for us.

**Comment by blandinw** [comment]
This is now fixed, thank you for your feedback and patience.
